### PR TITLE
Fixed #12885 -- joining along GenericRelation to child model

### PR DESCRIPTION
**Pull request by akaariai**

### CODE REVIEW

File: django/contrib/contenttypes/fields.py [hunk 1]
```diff
@@ -359,10 +359,43 @@ def resolve_related_fields(self):
         self.to_fields = [self.model._meta.pk.name]
         return [(self.remote_field.model._meta.get_field(self.object_id_field_name), self.model._meta.pk)]
 
+    def _get_path_info_with_parent(self):
+        """
+        Returns the path with joining through the parent model.
```
Comment (timgraham) [inline django/contrib/contenttypes/fields.py]: We are using PEP257 style for new docstrings "Return ..."
I'm having trouble understanding "with joining through the parent model" -- is it correct?

File: django/contrib/contenttypes/fields.py [hunk 2]
```diff
@@ -359,10 +359,43 @@ def resolve_related_fields(self):
...
+        """
+        # In a situation where you have inheritance chain ChildTag -> Tag,
+        # and Tagged defines the GenericForeignKey, and a TaggedItem model has a
```
Comment (timgraham) [inline django/contrib/contenttypes/fields.py]: did you mean "Tag defines"?

File: django/contrib/contenttypes/fields.py [hunk 3]
```diff
@@ -359,10 +359,43 @@ def resolve_related_fields(self):
...
+        # TaggedItem to Tag (as Tag.object_id == TaggedItem.pk), and next join
+        # from Tag to ChildTag (as that is where the relation is to). We do this
+        # by first generating a join to the parent model, then generating joins
+        # to the child models.
+        path = []
+        opts = self.remote_field.model._meta
+        parent_opts = opts.get_field(self.object_id_field_name).model._meta
+        target = parent_opts.pk
+        path.append(PathInfo(self.model._meta, parent_opts, (target,), self.remote_field, True, False))
+        # Collect joins needed for the parent -> child chain. This is easies to do
```
Comment (timgraham) [inline django/contrib/contenttypes/fields.py]: easiest*

File: tests/generic_relations/tests.py [hunk 4]
```diff
@@ -405,6 +405,13 @@ def test_generic_relation_to_inherited_child(self):
         # GenericRelations to models that use multi-table inheritance work.
         granite = ValuableRock.objects.create(name='granite', hardness=5)
         ValuableTaggedItem.objects.create(content_object=granite, tag="countertop", value=1)
+        self.assertEqual(ValuableRock.objects.filter(tags__value=1).count(), 1)
+        # We are generating a big inefficient query for tags__tag - we join first
```
Comment (timgraham) [inline tests/generic_relations/tests.py]: I'm not sure the purpose of this comment -- is it describing something that could be improved or just the fact that the implementation isn't very elegant (and probably can't be)?

**Comment by timgraham** [comment]
Feel free to respond to the questions and I'll make those fixes along with my edits. Thanks.

File: django/contrib/contenttypes/fields.py [hunk 5]
```diff
@@ -359,10 +359,43 @@ def resolve_related_fields(self):
...
+    def _get_path_info_with_parent(self):
+        """
+        Returns the path with joining through the parent model.
+        """
+        # In a situation where you have inheritance chain ChildTag -> Tag,
+        # and Tagged defines the GenericForeignKey, and a TaggedItem model has a
+        # GenericRelation to ChildTag, then we need to generate a join from
+        # TaggedItem to Tag (as Tag.object_id == TaggedItem.pk), and next join
+        # from Tag to ChildTag (as that is where the relation is to). We do this
+        # by first generating a join to the parent model, then generating joins
```
Comment (charettes) [inline django/contrib/contenttypes/fields.py]: "is pointing to"?

File: tests/generic_relations/tests.py (hunk 4 above)
Comment (akaariai) [inline tests/generic_relations/tests.py]: big inefficient -> bit inefficient. The query has one extra join we could technically remove. But likely we won't ever do that.

File: django/contrib/contenttypes/fields.py (hunk 1 above)
Comment (akaariai) [inline django/contrib/contenttypes/fields.py]: We return a path that has first a join to the parent model, then to the current model. The idea is that if you have a GFK defined on parent model, then we need to join the parent model first, then the child model.

File: django/contrib/contenttypes/fields.py (hunk 2 above)
Comment (akaariai) [inline django/contrib/contenttypes/fields.py]: Yes. The problem is that if you have TaggedItem and it has a relation to ChildTag, then currently (before this PR) a filter against the childtag joins directly from TaggedItem to ChildTag. But the needed fields for the join condition (object_id and content_type) are on the Tag model.

**Comment by timgraham** [comment]
merged in cd0ba8053dfdc81ade9a656dc8049de600638501, thanks!

**Pull request closed without merging**